libdnf/sack/query.cpp

```diff
@@ -769,6 +879,64 @@ Query::addFilter(HyNevra nevra, bool icase)
     return 0;
 }
 
+void
+Query::Impl::filterNevraStrict(int cmpType, const char **matches)
+{
+    Pool *pool = dnf_sack_get_pool(sack);
+    std::vector<NevraID> compareSet;
+    const unsigned nmatches = g_strv_length((gchar**)matches);
+    compareSet.reserve(nmatches);
+    NevraID nevraId;
+    for (unsigned int i = 0; i < nmatches; ++i) {
+        const char * nevraPattern = matches[i];
+        if (!nevraPattern)
+            throw std::runtime_error("Query can not accept NULL for STR match");
+        if (nevraId.parse(pool, nevraPattern)) {
+            compareSet.push_back(nevraId);
+        }
+    }
+    Map nevraResult;
+    map_init(&nevraResult, pool->nsolvables);
+
+    if (compareSet.size() > 1) {
+        std::sort(compareSet.begin(), compareSet.end(), nevraIDSorter);
+
+        Id id = -1;
+        while (true) {
+            id = result->next(id);
+            if (id == -1)
+                    break;
+            Solvable* s = pool_id2solvable(pool, id);
+            auto low = std::lower_bound(compareSet.begin(), compareSet.end(), *s,
+                                        nevraCompareLowerSolvable);
+            bool sameNEVRA = low->name == s->name && low->evr == s->evr && low->arch == s->arch;
+            if (low != compareSet.end() && sameNEVRA) {
+                MAPSET(&nevraResult, id);
+            }
+        }
+    } else {
+        for (auto & nevraId : compareSet) {
```

Comment:
Why `for`? There is one element in the `compareSet` vector maximally.
And if you use optimisation for `compareSet.empty()`, which I proposed above, then there will be exactly one element in the vector. Then you can replace `for` by `nevraId = compareSet[0];`.

Reply:
Ok, but in case HY_NOT it will not work.

Reply:
I think that it will work.